### ISSUE TITLE
Custom dynamic menus ?

### ISSUE THREAD

**Issue by butterw** [tsl0922/mpv-menu-plugin#28]
What is the preferred approach for custom dynamic menus, should they use a standalone user script or be added to dyn_menu.lua ?
They would make menu updates more frequent. ex: I would be interested in a shaders-on checked menu item.

**Comment by butterw** [comment]
I was trying to add a #@mute dynamic entry to the toplevel menu: **checked if mute.** 
A condition to update menu items vs submenus needs to be added to dyn_menu.lua (the entry was staying disabled without the else).

```-- handle #@mute menu item update
local function update_mute_menu(item)
    mp.observe_property('mute', 'bool', function(_, value)
		item.state = value and {'checked'} or {}
		menu_items_dirty = true
	end)
end

-- update dynamic menu item and handle submenu update
local function dyn_menu_update(item, keyword)
    if keyword == 'mute' then 
		update_mute_menu(item)
    else
    item.type = 'submenu'
[...]
```

**Comment by tsl0922** [comment]
> **NOTE:** be aware that `dyn_menu.lua` is conflict with other scripts that also update the `user-data/menu/items` property, it may be resolved in the later release.

You should add it in `dyn_menu.lua`.

**Comment by butterw** [comment]
Are you interested in adding checkbox entries such as  #@mute or #@shaders-on ("checked" if glsl-shaders is not empty) to default dyn_menu.lua ?
I could also see checkboxes being used inside a #@window-scale or #@speed submenu.

Update logic:
- when mpv starts: all dynamic menus need to be updated.
- when a new file is loaded: per file menus such as #@tracks need to be updated.
- when settings are changed (by the user):  #@mute or #@shaders-on need to be updated.

**Comment by tsl0922** [comment]
Well, I don't think it's worth to implement such logic with code, `cycle mute` already has OSD message shown, why not just show OSD message when switching?

Tracks and Chapters needs lua code because it's a dynamic list.

**Comment by butterw** [comment]
I agree that #@mute isn't indispensable and it may be better to display the mute status elsewhere (OSD status icon or term-status-msg rather than only osd message), but it was a simple example of a checked item, which is a feature of dynamic windows menus. #@shade (glsl-shaders) checked item is what I was really interested in, #@loop switch would be of interest to someone who toggles loop-file, etc. 
IMO whether new #@words are added or not, dyn_menu.lua needs to support custom dynamic menu items (not only submenus) with title or state updates based on user interaction. Then there is the case of menus with one checked value (ex: speed in youtube), where the checked state needs to updated (default value=1x, can be changed at runtime).

**Comment by butterw** [comment]
Here's what my shaders menu could look like in v2 ![mpv-shaders-menu](https://github.com/tsl0922/mpv-menu-plugin/assets/408915/2c0f9d75-f410-49cc-a6b1-aed387b602df), (I've only done the ON part so far) the list is a lot more readable with the checkmark for the toggled shader "color".
the --- in the first page can be used to dismiss the menu with left-click  (#23).

**Comment by tsl0922** [comment]
Glad you solved it 👍

The purpose of `dyn_menu.lua` is more to provide an example of how to write a script that updates the menu, I'm not going to add more none-common features to it, it is recommended to modify it yourself. Of course, you can share your scripts with others here

**Comment by butterw** [comment]
I see you've just added support for #@playlist, what new features do you intend to add to dyn_menu.lua ?

I would include a dynamic checkmark to mute, if I was adding mute to the top-level of a simple launcher such as:  
[https://github.com/tsl0922/mpv-menu-plugin/assets/408915/0f600fec-95cd-45f9-b472-d3c41541053d](url)
My understanding is that it is best to do all dynamic menu updates from a single point (dyn_menu.lua). Ideally I would like to move my customizations to a separate file to avoid having to maintain a fork of dyn_menu.lua (no idea if there is a way to do this with lua ?).

I think users may want to override: 
local function build_track_title(track, prefix, filename) -- track display format
local function dyn_menu_update(item, keyword) -- custom keywords, +checked_menu_items

**Comment by tsl0922** [comment]
You may just fork it, that's the reason why I moved the dynamic menu logic to lua.

**Comment by tsl0922** [comment]
I have added a syntax for checked menu item, it should meet your needs now.

**Comment by butterw** [comment]
It does work for booleans: mute, pause. But most mpv properties aren't so easy to work with.

According to debug-plugin loop is a flag when =no and a string when =inf. 

There are 2 possible values for empty glsl-shaders depending on how it was setup: empty table (it works in this case) or a table with an empty string (this is what I've been using and it fails here).

I've tried the reverse check on the number video-zoom (default value 0.0). It doesn't fully work. If I add 0.05 then do -0.05 the value is 7e-10 rather than 0, so zero is not checked. 
```
+ add video-zoom 0.05	#menu: View > Zoom > + in
- add video-zoom -0.05 #menu: View > Zoom > - out 
0 set video-zoom 0 #menu: View > Zoom > 0 #@video-zoom:check!
```

The default values for window-scale, speed is 1.0 not 0.
I want to be able to setup keybindings for individual window-scale values and this isn't possible when using a dynamic submenu.

Maybe a generic solution isn't possible for such cases ? In my custom script, I am using function check_if_condition(item, condition) with different conditions for each case.
 
Users typically want something that works with minimal effort, so I intend to stick to supported features only for my shader-pack config and this will still provide a big improvement over no menu. IMO forks when there are frequent updates to the main script ends up being too much trouble for end-users (ex: osc.lua forks).

**Comment by tsl0922** [comment]
It does support other types of values (read project README), but it's not easy to make it generic for all conditions.

That's why I recommend you to fork it.

**Comment by butterw** [comment]
- current version works on int values, ex: #@video-rotate:check!
- to get it working with float video-zoom I needed to modify as follows (some rounding occurs with math operations it seems):
```
if tp == 'number' then return math.abs(v)<1e-6 end
```
I modified to try to get window-scale checked for the value 1. The observed value is nil when window-scale is changed, so I needed
value = mp.get_property_number('window-scale')
- it works with loop-file, but not with loop.
